lib/system/freertos/mutex.h

```diff
@@ -16,24 +16,24 @@
 #ifndef __METAL_FREERTOS_MUTEX__H__
 #define __METAL_FREERTOS_MUTEX__H__
 
-#include <metal/atomic.h>
+#include <metal/assert.h>
+#include "FreeRTOS.h"
+#include "semphr.h"
+
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 typedef struct {
-	atomic_int v;
+	SemaphoreHandle_t m;
 } metal_mutex_t;
 
-#define METAL_MUTEX_UNLOCKED 0
-#define METAL_MUTEX_LOCKED   1
-
 /*
  * METAL_MUTEX_INIT - used for initializing an mutex element in a static struct
  * or global
  */
-#define METAL_MUTEX_INIT(m)	{ ATOMIC_VAR_INIT(METAL_MUTEX_UNLOCKED) }
+#define METAL_MUTEX_INIT(m) { NULL }
```

Comment:
This will likely create code that compiles, but does not work in some unexpected ways, since there will be a mutex that is uninitialized.
We should at least warn the user about this. For gcc, we could do something like:
#define METAL_MUTEX_INIT(m) _Pragma("GCC warning \"Static initialization of mutexes is unsupported.\");

Reply:
On second thought. It might be better to treat it as an error. s/warning/error/.

Reply:
That's a good point. 
Something that it makes the picture more complex is that is macro is used to initialize a structure on declaration, as done in the  METAL_MUTEX_DEFINE macro.

I had a look for each supported compiler  GCC, EWARN and ARMCC, 
- GCC support the warning message that can be changed in error
- EWARM (IAR) allows only to print a message or an error
- not possible to use _Pragma for print a warning.

il will send an extra patch on top

Reply:
This causes libmetal-test-static to fail intermittently and inconsistently on the Xilinx R5. I think we can keep this open and fix it, but this change shouldn't go into 2021.04 as is.

Reply:
> This causes libmetal-test-static to fail intermittently and inconsistently on the Xilinx R5. I think we can keep this open and fix it, but this change shouldn't go into 2021.04 as is.

Agree, not reasonnable to integrate it without understanding and fixing the issue.